pyflakes/test/test_other.py

```diff
@@ -13,6 +13,97 @@ class Test(TestCase):
     def test_duplicateArgs(self):
         self.flakes('def fu(bar, bar): pass', m.DuplicateArgument)
 
+    def test_duplicate_keys(self):
+        self.flakes("{'yes': 1, 'yes': 1}", m.DuplicateDictionaryKey)
+
+    def test_multiple_duplicate_keys(self):
+        self.flakes(
+            "{'yes': 1, 'yes': 1, 'no': 2, 'no': 3}",
+            m.DuplicateDictionaryKey, m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_in_function(self):
+        self.flakes('''
+        def f(thing):
+            pass
+        f({'yes': 1, 'yes': 4})
+        ''', m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_in_lambda(self):
+        self.flakes("lambda x: {(0,1): 1, (0,1): 1}",
+                    m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_tuples(self):
+        self.flakes("{(0,1): 1, (0,1): 1}", m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_ints(self):
+        self.flakes("{1: 1, 1: 1}", m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_bools(self):
+        if version_info < (3, 4):
+            expected = m.DuplicateVariableDictionaryKey
+        else:
+            expected = m.DuplicateDictionaryKey
+        self.flakes("{True: 1, True: 1}", expected)
+
+    def test_duplicate_keys_none(self):
+        if version_info < (3, 4):
+            expected = m.DuplicateVariableDictionaryKey
+        else:
+            expected = m.DuplicateDictionaryKey
+        self.flakes("{None: 1, None: 1}", expected)
+
+    def test_duplicate_variable_keys(self):
+        self.flakes('''
+        a = 1
+        {a: 2, a: 3}
+        ''', m.DuplicateVariableDictionaryKey)
+
+    def test_duplicate_key_float_and_int(self):
+        self.flakes('''
+        {1: 1, 1.0: 1}
+        ''', m.DuplicateDictionaryKey)
+
+    def test_no_duplicate_key_errors(self):
+        self.flakes('''
+        {'yes': 1, 'no': 1}
+        ''')
+
+    def test_no_duplicate_key_errors_func_call(self):
+        self.flakes('''
+        def test(thing):
+            pass
+        test({True: 1, None: 2, False: 1})
+        ''')
+
+    def test_no_duplicate_key_errors_bool_or_none(self):
+        self.flakes("{True: 1, None: 2, False: 1}")
+
+    def test_no_duplicate_key_errors_ints(self):
+        self.flakes('''
+        {1: 1, 2: 1}
+        ''')
+
+    def test_no_duplicate_key_errors_vars(self):
```

Comment:
I considered this, but I'm unsure. In most cases I would expect that variables will be being dynamically set, such that adding detection that actually looks up values where possible is a lot more effort than the reward it gains.

However, I suppose in cases of constants defined in the code it may well have some value so agreed that TODO is valid for this.